From 38e2ab8b3b846000977aa987b83725760c68a1ee Mon Sep 17 00:00:00 2001
From: Huang Chaofan <chaofan.huang@nxp.com>
Date: Mon, 25 Jun 2018 17:59:38 +0800
Subject: [PATCH 4091/5242] MLK-18679 VPU: fix the issue that Andriod has
 memory leak when call several times reqbufs

commit  2d177686c119e348b05c5d2bf482a6059c8dbcb8 from
https://source.codeaurora.org/external/imx/linux-imx.git

fix this issue by free the mbi buffer when request buffer count 0

Signed-off-by: Huang Chaofan <chaofan.huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/vpu-decoder-b0/vpu_b0.c |   18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/drivers/mxc/vpu-decoder-b0/vpu_b0.c b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
index 2d8a879..797b649 100644
--- a/drivers/mxc/vpu-decoder-b0/vpu_b0.c
+++ b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
@@ -610,9 +610,21 @@ static int v4l2_ioctl_reqbufs(struct file *file,
 	ret = vb2_reqbufs(&q_data->vb2_q, reqbuf);
 	if (!ret) {
 		if (reqbuf->type == V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE) {
-			for (i = 0; i < reqbuf->count; i++)
-				q_data->vb2_reqs[i].status = FRAME_ALLOC;
-			alloc_mbi_buffer(ctx, q_data, reqbuf->count);
+			if (reqbuf->count == 0) {
+				for (i = 0; i < MAX_MBI_NUM; i++)
+					if (ctx->mbi_dma_virt[i] != NULL) {
+						dma_free_coherent(&ctx->dev->plat_dev->dev,
+								ctx->mbi_size,
+								ctx->mbi_dma_virt[i],
+								ctx->mbi_dma_phy[i]
+								);
+						ctx->mbi_dma_virt[i] = NULL;
+					}
+			} else {
+				for (i = 0; i < reqbuf->count; i++)
+					q_data->vb2_reqs[i].status = FRAME_ALLOC;
+				alloc_mbi_buffer(ctx, q_data, reqbuf->count);
+			}
 		}
 	} else if (reqbuf->count != 0)
 		vpu_dbg(LVL_ERR, "error: %s() can't request (%d) buffer\n", __func__, reqbuf->count);
-- 
1.7.9.5

